Cards/CardBase.gd

```diff
@@ -42,8 +45,46 @@ func on_card_play(caster: Entity, base_target: Entity) -> void:
 	if caster is Player:
 		PlayerManager.player.get_energy_component().use_energy(self)	
 	
-	for effect_data: EffectData in card_effects_data:
-		var list_targets: Array[Entity] = effect_data.targeting_function.generate_target_list(base_target)
-		for current_target in list_targets:
-			effect_data.apply_effect_data(caster, current_target)
-	CardManager.on_card_action_finished.emit(self)
+	if _card_effects_queue.size() > 0:
+		push_error("Tried to play more card animations while some are already playing!")
+		return
+	
+	_card_effects_queue = card_effects_data.duplicate()
+	_handle_effects_queue(caster, base_target)
+
+
+func _handle_effects_queue(caster: Entity, base_target: Entity) -> void:
+	var card_effect: EffectData = _card_effects_queue[0]
+	var animation_data: CastAnimationData = card_effect.animation_data
+	var list_targets: Array[Entity] = card_effect.targeting_function.generate_target_list(base_target)
+	
+	var can_use_animation: bool = animation_data != null and\
+		animation_data.cast_animation_scene != null and\
+		animation_data.cast_position != null and\
+		!CardManager.disable_card_animations
+	
+	var created_cast_animation: CastAnimation = null
+	
+	if can_use_animation:
+		created_cast_animation = animation_data.cast_position.initialize_animation(animation_data.cast_animation_scene, list_targets)
+		
+		# Wait for animation hit to complete
+		await created_cast_animation.on_animation_hit_triggered
+	else:
+		push_warning("No animation set in effect data for card " + resource_path + ". Skipping animation.")
+	
+	_card_effects_queue.remove_at(0)
+	
+	# Apply effects to all targets
+	for current_target in list_targets:
+		card_effect.apply_effect_data(caster, current_target)
+	
+	# Wait for animation to complete
+	if created_cast_animation != null:
+		await created_cast_animation.on_animation_cast_complete
+	
+	# Handle next effect in the queue or finish casting
+	if _card_effects_queue.size() > 0:
+		_handle_effects_queue(caster, base_target)
```

Comment:
is there any specific reason for making this a recursion instead of a for loop ?

Reply:
Because we are using `await` in the above code, so this function will not run instantly. You can't use `await` inside a for loop as far as i know

Reply:
I see, i'm not too familiar with using the await